healthcheck/handler.go

```diff
@@ -0,0 +1,109 @@
+package healthcheck
+
+import (
+	"encoding/json"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/pkg/errors"
+	"net/http"
+	"time"
+)
+
+const (
+	StatusOK       = "OK"
+	StatusCritical = "CRITICAL"
+	StatusWarning  = "WARNING"
+)
+
+// Handler returns the current health check of the app
+func (hc HealthCheck) Handler(w http.ResponseWriter, req *http.Request) {
+	now := time.Now().UTC()
+	ctx := req.Context()
+
+	var checks []Check
+
+	for _, client := range hc.Clients {
+		if client.Check != nil {
+			checks = append(checks, *client.Check)
+		}
+	}
+
+	hc.Status = hc.getStatus()
+	hc.Uptime = now.Sub(hc.StartTime)
+		hc.Checks =    checks
+
+
+	b, err := json.Marshal(hc)
+
+	if err != nil {
+		log.ErrorCtx(ctx, errors.Wrap(err, "failed to marshal json"), log.Data{"error": err, "health_check_response": hc})
+		return
+	}
+
+	_, err = w.Write(b)
+	if err != nil {
+		log.ErrorCtx(ctx, errors.Wrap(err, "failed to write bytes for http response"), nil)
+		return
+	}
+}
+
+// isAppStartingUp returns true or false based on if each client has had results back from at least a single check
+func (hc HealthCheck) isAppStartingUp() bool {
+	for _, client := range hc.Clients {
+		if client.Check == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// getStatus returns a status as string as to the overall current apps health based on its dependent apps health
+func (hc HealthCheck) getStatus() string {
+	if hc.isAppStartingUp() {
+		return StatusWarning
+	}
+	return hc.isAppHealthy()
+}
+
+// isAppHealthy checks every clients check for their health then produces and returns a status for this apps health
+func (hc HealthCheck) isAppHealthy() string {
+	status := StatusOK
+	defer unlockAllClientChecks(hc.Clients)
```

Comment:
The godoc for mutex.Unlock:
`Unlock unlocks m. It is a run-time error if m is not locked on entry to Unlock`
An unexpected error like this would be bad so I think it might be cleaner to acquire each individual lock within a loop iteration to avoid this happening.

```
for _, client := range hc.Clients {
    client.mutex.Lock()
    defer client.mutex.Unlock()
    ....
}
```
We can guarantee then that we only ever call on unlock on a lock we have obtained.

Reply:
So I looked into doing it like that at first, but get prompts saying defers in loops can cause memory leaks ( which would also be bad if this was on all apps ) - although don't see how in our current situation it would cause a memory leak.

Reply:
I have instead separated this out into its own function so the locks are handled in a function call that is hit on every iteration